File: _Basic/windows-forms-introduction-with-visual-basic/8-windows-forms-introduction-with-visual-basic-m8-linq-exercise-files/CS/Data/Data/LinqSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Data
{
    public class LinqSource : ISource
    {
        public object GetCategories()
        {
            object result = null;

            using (NorthwindDataContext context = new NorthwindDataContext())
            {
                result = context.LinqCategories.ToList();
            }

            return result;
        }

        public object GetProducts(int categoryId)
        {
            object result = null;

            using (NorthwindDataContext context = new NorthwindDataContext())
            {
                var query = from p in context.LinqProducts
                            where p.CategoryID == categoryId
                            select p;

                result = query.ToList();
            }

            return result;
        }

        public bool AddProduct(Product product)
        {
            using (NorthwindDataContext context = new NorthwindDataContext())
            {
                LinqProduct linqProduct = new LinqProduct();
                linqProduct.ProductName = product.ProductName;
                linqProduct.SupplierID = 1;
                linqProduct.CategoryID = product.CategoryId;
                linqProduct.QuantityPerUnit = product.QuantityPerUnit;
                linqProduct.UnitPrice = product.UnitPrice;
                linqProduct.UnitsInStock = Convert.ToInt16(product.UnitsInStock);
                linqProduct.UnitsOnOrder = Convert.ToInt16(product.UnitsOnOrder);
                linqProduct.ReorderLevel = 0;
                linqProduct.Discontinued = product.Discontinued;

                context.LinqProducts.InsertOnSubmit(linqProduct);
                Save(context);
            }

            return true;
        }

        public bool DeleteProduct(int productId)
        {
            using (NorthwindDataContext context = new NorthwindDataContext())
            {
                var query = from p in context.LinqProducts
                            where p.ProductID == productId
                            select p;
                var linqProduct = query.Single();

                context.LinqProducts.DeleteOnSubmit(linqProduct);
                Save(context);
            }

            return true;
        }

        public void Save()
        {
            using (NorthwindDataContext context = new NorthwindDataContext())
            {
                Save(context);
            }
        }

        private void Save(NorthwindDataContext context)
        {
            try
            {
                context.SubmitChanges();
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
                throw;
            }
        }
    }
}

File: _Basic/windows-forms-introduction-with-visual-basic/8-windows-forms-introduction-with-visual-basic-m8-linq-exercise-files/VB/Data/Data/LinqSource.vb
Public Class LinqSource
    Implements ISource

    Public Function GetCategories() As Object Implements ISource.GetCategories
        Dim result As Object

        Using context As New NorthwindDataContext
            result = context.LinqCategories.ToList()
        End Using

        Return result
    End Function

    Public Function GetProducts(categoryId As Integer) As Object Implements ISource.GetProducts
        Dim result As Object

        Using context As New NorthwindDataContext
            Dim query = From p In context.LinqProducts
                        Where p.CategoryID = categoryId
                        Select p

            result = query.ToList()
        End Using

        Return result
    End Function

    Public Function AddProduct(product As Product) As Boolean Implements ISource.AddProduct
        Using context As New NorthwindDataContext
            Dim linqProduct As New LinqProduct()
            linqProduct.ProductName = product.ProductName
            linqProduct.SupplierID = 1
            linqProduct.CategoryID = product.CategoryId
            linqProduct.QuantityPerUnit = product.QuantityPerUnit
            linqProduct.UnitPrice = product.UnitPrice
            linqProduct.UnitsInStock = product.UnitsInStock
            linqProduct.UnitsOnOrder = product.UnitsOnOrder
            linqProduct.ReorderLevel = 0
            linqProduct.Discontinued = product.Discontinued

            context.LinqProducts.InsertOnSubmit(linqProduct)
            Save(context)
        End Using

        Return True
    End Function

    Public Function DeleteProduct(productId As Integer) As Boolean Implements ISource.DeleteProduct
        Using context As New NorthwindDataContext
            Dim query = From p In context.LinqProducts
                        Where p.ProductID = productId
                        Select p
            Dim linqProduct = query.Single()

            context.LinqProducts.DeleteOnSubmit(linqProduct)
            Save(context)
        End Using

        Return True
    End Function

    Public Sub Save() Implements ISource.Save
        Using context As New NorthwindDataContext
            Save(context)
        End Using
    End Sub

    Private Sub Save(context As NorthwindDataContext)
        Try
            context.SubmitChanges()
        Catch ex As Exception
            MsgBox(ex.Message)
            Throw
        End Try
    End Sub
End Class

